### PR TITLE
[JENKINS-35907] notify WorkspaceListener

### PR DESCRIPTION
**Pull request by ndeloof**
Downstream of https://github.com/jenkinsci/jenkins/pull/2410.

### CODE REVIEW

**Comment by jglick** [comment]
Resolve merge conflicts.

**Comment by jglick** [comment]
Do not forget to `mvn -f jenkins deploy` the dependency.

File: src/test/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStepTest.java [hunk 1]
```diff
@@ -24,11 +24,17 @@
 
 package org.jenkinsci.plugins.workflow.support.steps;
 
+import hudson.FilePath;
+import hudson.model.AbstractBuild;
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStepTest.java]: :ant: unused import

File: src/test/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStepTest.java [hunk 2]
```diff
@@ -24,11 +24,17 @@
...
+import hudson.model.BuildListener;
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStepTest.java]: :ant: ditto

File: src/test/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStepTest.java [hunk 3]
```diff
@@ -429,4 +436,30 @@ private void startJnlpProc() throws Exception {
         });
     }
 
+    private static class MyWorkspaceListener extends WorkspaceListener {
+        public boolean notified = false;
+
+        @Override
+        public void beforeUse(Run r, FilePath workspace, TaskListener listener) {
+            notified = true;
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStepTest.java]: Better to print something to `listener` and then use `JenkinsRule.assertLogContains`.

File: src/test/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStepTest.java [hunk 4]
```diff
@@ -429,4 +436,30 @@ private void startJnlpProc() throws Exception {
...
+        }
+    }
+    @TestExtension("callWorkspaceListener")
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStepTest.java]: :ant: use `@TestExtension` on the class, which must then be `public`.

File: src/test/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStepTest.java [hunk 5]
```diff
@@ -429,4 +436,30 @@ private void startJnlpProc() throws Exception {
...
+    public final static MyWorkspaceListener workspaceListener = new MyWorkspaceListener();
+
+
+    @Issue("JENKINS-35907")
+    @Test public void callWorkspaceListener() throws Exception {
+        story.addStep(new Statement() {
+            @Override
+            public void evaluate() throws Throwable {
+                WorkflowJob p = story.j.jenkins.createProject(WorkflowJob.class, "callWorkspaceListener");
+                p.setDefinition(new CpsFlowDefinition("node() { sh 'echo hello' }"));
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStepTest.java]: :ant: see nearby tests for the Windows equivalent, e.g.

```groovy
if (isUnix()) {sh 'echo hello'} else {bat 'echo hello'}
```

File: src/test/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStepTest.java [hunk 6]
```diff
@@ -429,4 +436,30 @@ private void startJnlpProc() throws Exception {
...
+                p.save();
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStepTest.java]: unnecessary BTW

File: src/test/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStepTest.java [hunk 7]
```diff
@@ -429,4 +436,30 @@ private void startJnlpProc() throws Exception {
...
+                WorkflowRun b = p.scheduleBuild2(0).waitForStart();
+                story.j.waitForCompletion(b);
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStepTest.java]:
```java
WorkflowRun b = r.assertBuildStatusSuccess(p.scheduleBuild2(0));
```

**Comment by jglick** [comment]
I suppose this is also supposed to be @reviewbybees.

File: src/test/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStepTest.java [hunk 8]
```diff
@@ -429,4 +433,26 @@ private void startJnlpProc() throws Exception {
...
+
+    @Issue("JENKINS-35907")
+    @Test public void callWorkspaceListener() throws Exception {
+        story.addStep(new Statement() {
+            @Override
+            public void evaluate() throws Throwable {
+                WorkflowJob p = story.j.jenkins.createProject(WorkflowJob.class, "callWorkspaceListener");
+                p.setDefinition(new CpsFlowDefinition("node() { echo 'hello' }"));
+                WorkflowRun b = story.j.assertBuildStatusSuccess(p.scheduleBuild2(0));
+                story.j.waitForCompletion(b);
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStepTest.java]: :ant: redundant

**Comment by jglick** [comment]
Still needs a merge with `master`.

**Comment by jglick** [comment]
Need to redeploy a core snapshot I think.

**Comment by dwnusbaum** [comment]
Looks like the upstream PR was closed without being merged, so I am going to close this one as well. Please feel free to reopen if the PRs are brought up to date again.

**Pull request closed without merging**